File: src/test/java/com/hedvig/botService/chat/ConversationTest.java
package com.hedvig.botService.chat;

import com.hedvig.botService.enteties.MemberChat;
import com.hedvig.botService.enteties.UserContext;
import com.hedvig.botService.enteties.message.*;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.*;
import org.mockito.runners.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.Arrays;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.BDDMockito.then;

@RunWith(MockitoJUnitRunner.class)
public class ConversationTest {

    public static final String TESTMESSAGE_ID = "testmessage";

    @Mock(answer = Answers.CALLS_REAL_METHODS)
    Conversation sut;

    UserContext uc;

    @Spy
    MemberChat mc;

    @Before
    public void setup() {
        uc = new UserContext();
        uc.setMemberChat(mc);
    }



    @Captor
    ArgumentCaptor<Message> messageCaptor;


    @Test
    public void addToChat_renders_selectLink() throws Exception {
        //Arrange
        uc.putUserData("{TEST}", "localhost");

        Message m = createSingleSelectMessage("En förklarande text",
                new SelectLink("Länk text", "selected.value", null, "bankid:///{TEST}/text", "http://{TEST}/text", false));

        //ACT
        sut.addToChat(m, uc);

        //Assert
        then(mc).should().addToHistory(messageCaptor.capture());
        MessageBodySingleSelect body = (MessageBodySingleSelect) messageCaptor.getValue().body;

        SelectLink link = (SelectLink) body.choices.get(0);
        assertThat(link.appUrl).isEqualTo("bankid:///localhost/text");
        assertThat(link.webUrl).isEqualTo("http://localhost/text");

    }

    @Test
    public void addToChat_renders_message() throws Exception {
        //Arrange
        uc.putUserData("{REPLACE_THIS}", "kort");

        Message m = createTextMessage("En förklarande {REPLACE_THIS} text");

        //ACT
        sut.addToChat(m, uc);

        //Assert
        then(mc).should().addToHistory(messageCaptor.capture());
        MessageBody body =  messageCaptor.getValue().body;

        assertThat(body.text).isEqualTo("En förklarande kort text");

    }

    private Message createSingleSelectMessage(final String text, final SelectItem... items) {
        Arrays.asList(items);

        return createMessage(new MessageBodySingleSelect(text, Arrays.asList(items)));
    }

    private Message createTextMessage(final String text) {
        return createMessage(new MessageBodyText(text));
    }

    private Message createMessage(MessageBody body) {
        Message m = new Message();
        m.id = TESTMESSAGE_ID;
        m.globalId = 1;
        m.header = new MessageHeader();
        m.header.fromId = -1L;
        m.header.messageId = 1;
        m.body = body;
        return m;
    }

}

File: src/test/java/com/hedvig/botService/chat/ConversationTest.kt
package com.hedvig.botService.chat

import com.hedvig.botService.enteties.MemberChat
import com.hedvig.botService.enteties.UserContext
import com.hedvig.botService.enteties.message.*
import com.hedvig.botService.testHelpers.MessageHelpers.createSingleSelectMessage
import com.hedvig.botService.testHelpers.MessageHelpers.createTextMessage
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.*
import org.mockito.BDDMockito.then
import org.mockito.runners.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class ConversationTest {

  @Mock(answer = Answers.CALLS_REAL_METHODS)
  internal lateinit var sut: Conversation

  private lateinit var uc: UserContext

  @Spy
  internal var mc: MemberChat? = null
  @Captor
  internal var messageCaptor: ArgumentCaptor<Message>? = null

  @Before
  fun setup() {
    uc = UserContext()
    uc.memberChat = mc
  }

  @Test
  @Throws(Exception::class)
  fun addToChat_renders_selectLink() {
    // Arrange
    uc.putUserData("{TEST}", "localhost")

    val m = createSingleSelectMessage(
      "En förklarande text",
      SelectLink(
        "Länk text",
        "selected.value",
        null,
        "bankid:///{TEST}/text",
        "http://{TEST}/text",
        false))

    // ACT
    sut.addToChat(m, uc)

    // Assert
    then<MemberChat>(mc).should().addToHistory(messageCaptor!!.capture())
    val body = messageCaptor!!.value.body as MessageBodySingleSelect

    val link = body.choices[0] as SelectLink
    assertThat(link.appUrl).isEqualTo("bankid:///localhost/text")
    assertThat(link.webUrl).isEqualTo("http://localhost/text")
  }

  @Test
  @Throws(Exception::class)
  fun addToChat_renders_message() {
    // Arrange
    uc.putUserData("{REPLACE_THIS}", "kort")

    val m = createTextMessage("En förklarande {REPLACE_THIS} text")

    // ACT
    sut.addToChat(m, uc)

    // Assert
    then<MemberChat>(mc).should().addToHistory(messageCaptor!!.capture())
    val body = messageCaptor!!.value.body

    assertThat(body.text).isEqualTo("En förklarande kort text")
  }

  @Test
  fun `receiveMessage_withRegisteredSingleSelectCallback_callsCallback`() {
    var called = false

    val testClass = makeConversation {
      this.createChatMessage(
        "message.id", WrappedMessage(
        MessageBodySingleSelect(
          "hej", listOf(
          SelectItem(false, "Text", "value")))
      ) { body, usercontext, _ ->
        called = true
        ""
      })
    }


    testClass.receiveMessage(uc, makeMessage("message.id", MessageBodySingleSelect("", listOf())))

    assertThat(called).isTrue()

  }

  @Test
  fun receiveMessage_withRegisteredMessageBodyTextCallback_callsCallback() {
    var called = false

    val testClass = makeConversation {
      this.createChatMessage(
        "message.id", WrappedMessage(
        MessageBodyText("hej"),
        callback = { _, _, _ ->
          called = true
          ""
        }))
    }


    testClass.receiveMessage(uc, makeMessage("message.id", MessageBodyText("")))

    assertThat(called).isTrue()

  }

  fun makeMessage(id: String, body: MessageBody): Message {
    val m = Message()
    m.id = id
    m.body = body
    return m
  }


  fun makeConversation(constructor: Conversation.(Unit) -> Unit): Conversation {
    return object : Conversation() {
      override fun getSelectItemsForAnswer(uc: UserContext): List<SelectItem> {
        return listOf()
      }

      override fun canAcceptAnswerToQuestion(uc: UserContext): Boolean {
        return false
      }

      override fun handleMessage(userContext: UserContext, m: Message) {

      }

      override fun init(userContext: UserContext) {
      }

      override fun init(userContext: UserContext, startMessage: String) {
      }

      init {
        constructor.invoke(this, Unit)
      }
    }
  }

  companion object {

    @JvmField
    val TESTMESSAGE_ID = "testmessage"
  }
}
